src/mystery_gift.c

```diff
@@ -0,0 +1,293 @@
+}
+
+BOOL SaveMysteryGiftI_TagIsValid(MYSTERY_GIFT_SAVE* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_MYSTERY_GIFTS);
+    
+    if (MysteryGiftTagIsValid(mg->gifts[index].tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_CardTagIsValid(MYSTERY_GIFT_SAVE* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_WONDER_CARDS);
+    
+    if (MysteryGiftTagIsValid(mg->cards[index].tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_SpecialCardTagIsValid(MYSTERY_GIFT_SAVE* mg) {
+    if (MysteryGiftTagIsValid(mg->specialWonderCard.tag)) {
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+BOOL SaveMysteryGift_HasAnyCard(MYSTERY_GIFT_SAVE* mg) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_WONDER_CARDS; ++i) {
+        if (SaveMysteryGift_CardTagIsValid(mg, i) == TRUE) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_HasAnyGift(MYSTERY_GIFT_SAVE* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+void SaveMysteryGift_SetReceivedByCardId(MYSTERY_GIFT_SAVE* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            SaveMysteryGiftI_SetReceived(mg, i);
+            break;
+        }
+    }
+}
+
+BOOL SaveMysteryGift_ReceivedFlagTest(MYSTERY_GIFT_SAVE* mg, int a1) {
+    GF_ASSERT(a1 < 2048);
+    return mg->filler_000[a1 / 8] & (1 << (a1 & 7)) ? TRUE : FALSE;
+}
+
+void SaveMysteryGift_ReceivedFlagSet(MYSTERY_GIFT_SAVE* mg, int a1) {
+    GF_ASSERT(a1 < 2048);
+    mg->filler_000[a1 / 8] |= (1 << (a1 & 7));
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+void SaveMysteryGift_ReceivedFlagClear(MYSTERY_GIFT_SAVE* mg, int a1) {
+    u8 mask = ~(1 << (a1 & 7));
+    GF_ASSERT(a1 < 2048);
+    mg->filler_000[a1 / 8] &= mask;
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+BOOL SaveMysteryGift_TestFlagx7FF(MYSTERY_GIFT_SAVE* mg) {
+    return SaveMysteryGift_ReceivedFlagTest(mg, 0x7FF);
+}
+
+void SaveMysteryGift_SetFlagx7FF(MYSTERY_GIFT_SAVE* mg) {
+    SaveMysteryGift_ReceivedFlagSet(mg, 0x7FF);
+}
+
+static MYSTERY_GIFT_SAVE* sMysteryGiftData;
+
+void GetStaticPointerToSaveMysteryGift(SaveData* saveData) {
+    if (sMysteryGiftData == NULL) {
+        sMysteryGiftData = Save_MysteryGift_Get(saveData);
+    }
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+void DeleteStaticPointerToMysteryGift(void) {
+    if (sMysteryGiftData != NULL) {
+        sMysteryGiftData = NULL;
+    }
+}
+
+int GetFirstQueuedMysteryGiftIdx(void) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (SaveMysteryGiftI_TagIsValid(sMysteryGiftData, i) == TRUE) {
```

Comment:
What's the difference between SaveMysteryGift and SaveMysteryGiftI?

Reply:
The I stands for index.